source/extensions/transport_sockets/alts/tsi_handshaker.cc

```diff
@@ -0,0 +1,72 @@
+#include "extensions/transport_sockets/alts/tsi_handshaker.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+void TsiHandshaker::onNextDone(tsi_result status, void* user_data,
+                               const unsigned char* bytes_to_send, size_t bytes_to_send_size,
+                               tsi_handshaker_result* handshaker_result) {
+  TsiHandshaker* handshaker = static_cast<TsiHandshaker*>(user_data);
+
+  Buffer::InstancePtr to_send = std::make_unique<Buffer::OwnedImpl>();
+  if (bytes_to_send_size > 0) {
+    to_send->add(bytes_to_send, bytes_to_send_size);
+  }
+
+  auto next_result =
+      new TsiHandshakerCallbacks::NextResult{status, std::move(to_send), {handshaker_result}};
+
+  handshaker->dispatcher_.post([handshaker, next_result]() {
```

Comment:
Is there any reason why callbacks can't be called in current call stack?

Reply:
When `tsi_handshaker_next` returned `TSI_ASYNC`, this method will be called in a different thread owned by TSI, we post back the callback to the original thread to follow envoy thread model, so in the callback it doesn't need to lock buffers, etc.